lib/jxl/modular/encoding/context_predict.h

```diff
@@ -334,21 +334,20 @@ constexpr size_t kGradientProp = 9;
 // Clamps gradient to the min/max of n, w (and l, implicitly).
 static JXL_INLINE int32_t ClampedGradient(const int32_t n, const int32_t w,
                                           const int32_t l) {
-  const int32_t m = std::min(n, w);
-  const int32_t M = std::max(n, w);
-  // The end result of this operation doesn't overflow or underflow if the
-  // result is between m and M, but the intermediate value may overflow, so we
-  // do the intermediate operations in uint32_t and check later if we had an
-  // overflow or underflow condition comparing m, M and l directly.
-  // grad = M + m - l = n + w - l
-  const int32_t grad =
-      static_cast<int32_t>(static_cast<uint32_t>(n) + static_cast<uint32_t>(w) -
-                           static_cast<uint32_t>(l));
-  // We use two sets of ternary operators to force the evaluation of them in
-  // any case, allowing the compiler to avoid branches and use cmovl/cmovg in
-  // x86.
-  const int32_t grad_clamp_M = (l < m) ? M : grad;
-  return (l > M) ? m : grad_clamp_M;
+  const int32_t ab =
```

Comment:
A bunch of comments here would be very appreciated :)

Reply:
Added a bunch of them to help explain / prove why this actually works.

Reply:
Maybe we could also add a test that checks a bunch of cases vs the naive impl?

Reply:
When I was trying to come up with this somewhat less-branchy reformulation, I always did it by just brute-force checking, which is much easier and less error-prone than proving correctness :)
Something like this:
```C
for (int A = -300; A < 300; A++)
for (int B = -300; B < 300; B++)
for (int C = -300; C < 300; C++) {
    int a = ClampedGradient(A,B,C);
    int b = OtherClampedGradient(A,B,C);
    if (a != b) printf("NOT OK: %i %i %i -> %i != %i\n",A,B,C,a,b);
}
```

Something like that could be added as a unit test too but I'm not sure if that is very useful.

Reply:
I think it would be - especially for randomly weird systems...

Reply:
Fair enough, added a test that compares random combinations to the result of a naive implementation.

With values that cause int32 overflows in the intermediate results, there is a difference, so the test only tests values in the range [-2^29.. 2^29) — i.e. int30. Since libjxl cannot deal with integer bit depths > 24 anyway (since things have to be representable in binary32 floats), that's probably fine.

Lossless binary32 float encoding (reinterpreted as int32) still works, but the input does have to be in the [0.f,1.f) range (which maps to a subset of [0, 2^30) when reinterpreting as integer) or else weird overflows will happen. But that was already the case with the previous implementation too, I suppose.